### PR TITLE
misc: Migrate away from versionize

### LINKED ISSUE #6370: Consider dropping versionize
Firecracker have dropped it - https://github.com/firecracker-microvm/firecracker/pull/4230 and we have never taken advantage of the versioning in Cloud Hypervisor. I think we should just serde instead and take manual steps if we need to allow cross version migration (e.g. temporary custom logic to use multiple versions of the same field.) If we use json for the encoding we can take advantage of empty fields (and thus use default values) - or we could rewrite the state data upon migration / snapshot restore if necessary.

This would break live migration but due to various fixes live migration is not supported from v38.x - if we make a prompt decision we could enable this before the next release and take advantage of the existing breakage.

### PR DESCRIPTION
**Pull request by rbradford**
Replace with serde instead.

Fixes: #6370

Signed-off-by: Rob Bradford <rbradford@rivosinc.com>

### CODE REVIEW

**Review by likebreath — DISMISSED** [review]

**Review by up2wing — APPROVED** [review]

File: vm-migration/src/lib.rs [hunk 1]
```diff
@@ -106,23 +88,10 @@ impl SnapshotData {
...
-    where
-        T: Versionize + VersionMapped,
-    {
-        let mut data = Vec::new();
-        state
-            .serialize(&mut data, &T::version_map(), VMM_VERSION)
-            .map_err(|e| MigratableError::Snapshot(anyhow!("Error serialising: {}", e)))?;
-
-        Ok(SnapshotData(data))
+        Ok(SnapshotData { version: 1, state })
```
[the remark this answers is not in the document]
Comment (rbradford) [reply vm-migration/src/lib.rs:94]: > Nit: I think it is better to define a constant for the `snapshot_version`.
> 
> Also, I am thinking we may want to provide both `snapshot_major_version` and `snapshot_minor_version`. With this, we can make it clear that a major version bump indicates breaking live upgrade, and we can put checks in place to explicitly abort live upgrades from incompatible snapshots. Meanwhile, minor version bumps can be used for potential manual fix on live upgrades during a transitioning period of time.

I wanted to make the version semantic and separate from the the CH version - since it's more about the the contents changing. Adaption to the version would need to be local e.g. in the place that uses the snapshot data - it therefore would be best to know about the version.

Also I didn't want to use the CH version as that is something we would have to update on release - ideally we would never use this version field as we would agree to not break the updates any further.

I only chose `1` vs `0` as the value so that it would not be the default value for u64.

File: vm-migration/src/lib.rs [hunk 2]
```diff
@@ -139,6 +108,10 @@ impl SnapshotData {
 /// the actual device snapshot data.
 #[derive(Clone, Default, Deserialize, Serialize)]
 pub struct Snapshot {
+    /// Version of this snapshot: can be used for transitions in the future if
+    /// necessary
+    pub version: u64,
+
```
[the remark this answers is not in the document]
Comment (rbradford) [reply vm-migration/src/lib.rs:114]: I think I covered it in the previous  comment - but let me try again.

If say - we need to handle an upgrade that needs the version in the virtio-net code - it would need to know that the version of the virtio-net state has changed from version `x` to version `x+1` and adapt appropriately. 

I think that having version on `Snapshot` is useful in case we want to change the layout of the snapshot struct itself..

File: vm-migration/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (likebreath) [reply vm-migration/src/lib.rs:94]: Maybe I was not clear. I totally agree the idea of not using CH versions as the snapshot/snapshotdata versions. My comment was about defining a constant for it for better readability and using both major and minor versions for the case I mentioned above (mostly for explicitly aborting live-upgrades across breaking versions).

File: vm-migration/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (likebreath) [reply vm-migration/src/lib.rs:114]: > I think I covered it in the previous comment - but let me try again.
> 
> If say - we need to handle an upgrade that needs the version in the virtio-net code - it would need to know that the version of the virtio-net state has changed from version `x` to version `x+1` and adapt appropriately.
> 

My main point is, assuming the version of `SnapshotData` and `Snapshot` will always stay consistent (e.g. the same), it won't need to be included in both places, say we can always access the version information from the `Snapshot` struct via its `SnapshotData` field. 

> I think that having version on `Snapshot` is useful in case we want to change the layout of the snapshot struct itself..

For such case, I agree having separate version on `Snapshot` itself could be helpful. I'd say such kind of changes would be quite dramatic (indicating major refactoring on the way how we do snapshot/restore). I am not sure if we will be able to keep the live-upgrade in that case even with the version information.

[the remark this answers is not in the document]
Comment (rbradford) [reply vm-migration/src/lib.rs:114]: I envisioned the version being local to the thing that changed - that's why I gave the particular example of virtio-net.

The version for `SnapshotData` would change if the related `*State` struct changes e.g. if the `NetState` struct changed in a way that needs transition when then the code turning the `SnapshotData` into the `NetState` would need to handle that.

Perhaps the right approach is to add a `version` field to every `State` struct - that would make it more obvious? WDYT?

[the remark this answers is not in the document]
Comment (likebreath) [reply vm-migration/src/lib.rs:114]: Oh, I see. If that's the purpose of the version information you want, adding it to every `*State` struct is more obvious and better. 

But I don't see how such version information would be helpful for transitions when there are changes to the `*State` struct. Let's say the `NetState` is changed with a new field, and we give it a new version. How would the version bump within the `NetState` struct help with not breaking the live upgrade for the transition time?

[the remark this answers is not in the document]
Comment (rbradford) [reply vm-migration/src/lib.rs:114]: I've come to the same realisation that having it in the `*State` struct is too late - by that time it's already been deserialized from the `SnapshotData` - so we need the version in the `SnapshotData` so that when deserializing the restore logic can make the correct choice (e.g. deserializing to a different version of the struct or treating it as a dictionary and manually accessing the values for a transition.)

I'm going to take the version stuff out from this PR as I think it needs to be considered carefully and since we ant to expedite this change that seems the best plan.

**Comment by rbradford** [comment]
@likebreath I have removed the version fields - we can try a solution for that when we've had more time to discuss and explore options.

File: vm-migration/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (up2wing) [reply vm-migration/src/lib.rs]: > I don't see how such version information would be helpful for transitions when there are changes to the `*State` struct.

Agree.
If the `*State` struct change, the Restore will report an error such like this:
Error running command: Server responded with an error: InternalServerError: ApiError(VmRestore(Restore(Restore(Could not create devices RestoreGetState(Restore(Error deserialising: _net1-section A deserialization error occured: Io(Error { kind: UnexpectedEof, message: "failed to fill whole buffer" })))))))

[the remark this answers is not in the document]
Comment (rbradford) [reply vm-migration/src/lib.rs]: We have to handle it manually e.g. create if version == 1 deserialize to a e.g. `NetStateV1` struct and copy the values out where appropriate - or if necessary use fields with a serde default - it can be handled on a case by ase basis depending on what is needed for the transition.

**Review by likebreath — APPROVED** [review]
LGTM. Thanks for looking into this. It definitely helps with maintenance of live upgrade/migration in a long run. 

@rbradford I wonder do you know is there any performance differences between `to_string` vs `to_vec` from `serde_json`, say the speed and the size of serialized data?

/cc @up2wing

**Comment by rbradford** [comment]
> LGTM. Thanks for looking into this. It definitely helps with maintenance of live upgrade/migration in a long run.
> 
> @rbradford I wonder do you know is there any performance differences between `to_string` vs `to_vec` from `serde_json`, say the speed and the size of serialized data?

The string encoding might be marginally faster as it isn't encode as JSON array of each separate byte.  The state saving/loading is minor in comparison to the cost of the saving/loading the RAM!
